File: apps/api/src/security/file_validation.py
"""
Secure file validation utilities.
Blocks SVG files entirely to prevent XSS attacks (CWE-79).
Validates file types and content to prevent unrestricted uploads (CWE-434).
"""

import re
from typing import List, Optional, Tuple

from fastapi import HTTPException, UploadFile


def validate_image_content(content: bytes) -> bool:
    """Validate image content using magic bytes."""
    if len(content) < 12:
        return False

    # Check common image format magic bytes
    magic_bytes = content[:12]

    # JPEG: FF D8 FF
    if magic_bytes.startswith(b"\xff\xd8\xff"):
        return True

    # PNG: 89 50 4E 47 0D 0A 1A 0A
    if magic_bytes.startswith(b"\x89PNG\r\n\x1a\n"):
        return True

    # GIF: GIF87a or GIF89a
    if magic_bytes.startswith((b"GIF87a", b"GIF89a")):
        return True

    # AVIF: ftyp..avif / avis
    if magic_bytes[4:8] == b"ftyp" and magic_bytes[8:12] in {b"avif", b"avis"}:
        return True

    # WebP: RIFF....WEBP
    return bool(magic_bytes.startswith(b"RIFF") and b"WEBP" in content[:16])


def validate_audio_content(content: bytes) -> bool:
    """Validate audio content using magic bytes."""
    if len(content) < 12:
        return False

    magic_bytes = content[:12]

    # MP3: ID3 marker or frame header
    if magic_bytes.startswith(b"ID3") or magic_bytes[0] == 0xFF:
        return True

    # WAV: RIFF....WAVE
    if magic_bytes.startswith(b"RIFF") and content[8:12] == b"WAVE":
        return True

    # OGG: OggS header
    if magic_bytes.startswith(b"OggS"):
        return True

    # M4A: MP4 container with audio brand
    return magic_bytes[4:8] == b"ftyp" and magic_bytes[8:12] in {
        b"M4A ",
        b"isom",
        b"mp42",
        b"mp41",
    }


def validate_video_content(content: bytes) -> bool:
    """Validate video content using magic bytes."""
    if len(content) < 12:
        return False

    magic_bytes = content[:12]

    # MP4 / MOV: starts with specific ftyp box signatures
    if magic_bytes[4:8] == b"ftyp" and (
        b"mp4" in magic_bytes[8:12]
        or b"M4V" in magic_bytes[8:12]
        or b"isom" in magic_bytes[8:12]
        or magic_bytes[8:12] == b"qt  "
    ):
        return True

    # AVI: RIFF....AVI
    if magic_bytes.startswith(b"RIFF") and content[8:12] == b"AVI ":
        return True

    # FLV: FLV header
    if magic_bytes.startswith(b"FLV"):
        return True

    # WebM / MKV: EBML header
    return bool(magic_bytes.startswith(b"\x1aE\xdf\xa3"))


def validate_document_content(content: bytes) -> bool:
    """Validate document and archive content using magic bytes."""
    if len(content) < 4:
        return False

    # PDF: %PDF-
    if content.startswith(b"%PDF-"):
        return True

    # PPTX / DOCX / ZIP: ZIP container header
    if content.startswith((b"PK\x03\x04", b"PK\x05\x06", b"PK\x07\x08")):
        return True

    # VTT: WEBVTT
    if content.startswith(b"WEBVTT"):
        return True

    # SRT: plain text subtitle
    try:
        decoded = content.decode("utf-8-sig")
        return bool(decoded.strip())
    except UnicodeDecodeError:
        return False


# File type configurations
FILE_TYPES = {
    "image": {
        "extensions": [".jpg", ".jpeg", ".png", ".gif", ".webp", ".avif"],
        "mime_types": [
            "image/jpeg",
            "image/png",
            "image/gif",
            "image/webp",
            "image/avif",
        ],
        "max_size": 10 * 1024 * 1024,  # 10MB
        "validator": validate_image_content,
    },
    "video": {
        "extensions": [".mp4", ".webm", ".mkv", ".mov", ".avi", ".flv"],
        "mime_types": [
            "video/mp4",
            "video/webm",
            "video/x-matroska",
            "video/quicktime",
            "video/x-msvideo",
            "video/x-flv",
        ],
        "max_size": 1000 * 1024 * 1024,  # 1000MB
        "validator": validate_video_content,
    },
    "audio": {
        "extensions": [".mp3", ".wav", ".ogg", ".m4a", ".opus", ".oga"],
        "mime_types": [
            "audio/mpeg",
            "audio/wav",
            "audio/x-wav",
            "audio/ogg",
            "audio/opus",
            "audio/mp4",
            "audio/x-m4a",
        ],
        "max_size": 100 * 1024 * 1024,  # 100MB
        "validator": validate_audio_content,
    },
    "document": {
        "extensions": [".pdf", ".pptx", ".docx", ".zip", ".srt", ".vtt", ".txt"],
        "mime_types": [
            "application/pdf",
            "application/vnd.openxmlformats-officedocument.presentationml.presentation",
            "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            "application/zip",
            "application/x-zip-compressed",
            "text/vtt",
            "text/plain",
            "application/octet-stream",
        ],
        "max_size": 100 * 1024 * 1024,  # 100MB
        "validator": validate_document_content,
    },
}


def validate_upload(
    file: UploadFile, allowed_types: list[str], max_size: int | None = None
) -> tuple[str, bytes]:
    """
    Validate uploaded file for security and type compliance.

    Args:
        file: The uploaded file
        allowed_types: List of allowed file types ('image', 'video', 'audio', 'document')
        max_size: Maximum file size in bytes (auto-determined if None)

    Returns:
        Tuple of (mime_type, file_content)

    Raises:
        HTTPException: If validation fails
    """
    if not file or not file.filename:
        raise HTTPException(status_code=400, detail="No file provided")

    # Read file content once
    content = file.file.read()
    file.file.seek(0)

    # Get file extension and block SVG explicitly
    ext = "." + file.filename.split(".")[-1].lower()
    if ext == ".svg":
        raise HTTPException(
            status_code=415, detail="SVG files are not allowed for security reasons"
        )

    # Find matching file type configuration
    config = None
    for file_type in allowed_types:
        if file_type in FILE_TYPES and ext in FILE_TYPES[file_type]["extensions"]:
            config = FILE_TYPES[file_type]
            break

    if not config:
        allowed_exts = [
            ext
            for t in allowed_types
            for ext in FILE_TYPES.get(t, {}).get("extensions", [])
        ]
        raise HTTPException(
            status_code=415, detail=f"File type not allowed. Allowed: {allowed_exts}"
        )

    # Check file size
    size_limit = max_size or config["max_size"]
    if len(content) > size_limit:
        raise HTTPException(
            status_code=413,
            detail=f"File too large ({len(content) / 1024 / 1024:.1f}MB > {size_limit / 1024 / 1024:.1f}MB)",
        )

    # Validate file content
    if not config["validator"](content):
        raise HTTPException(
            status_code=415, detail="File appears to be corrupted or invalid"
        )

    return file.content_type, content


def get_safe_filename(original_filename: str, prefix: str = "") -> str:
    """Generate a safe filename with UUID and validated extension."""
    if not original_filename:
        return f"{prefix}.bin"

    ext = original_filename.rsplit(".", maxsplit=1)[-1].lower()
    # Only allow safe alphanumeric extensions
    if re.match(r"^[a-zA-Z0-9]+$", ext):
        return f"{prefix}.{ext}"

    return f"{prefix}.bin"

File: apps/web/src/lib/file-validation.ts
/**
 * Frontend file validation utilities
 * Provides consistent validation across all upload components
 */

// File type configurations (matches backend)
export const FILE_TYPES = {
  image: {
    extensions: ['.jpg', '.jpeg', '.png', '.gif', '.webp', '.avif'],
    mimeTypes: ['image/jpeg', 'image/png', 'image/gif', 'image/webp', 'image/avif'],
    maxSize: 20 * 1024 * 1024, // 20MB
  },
  video: {
    extensions: ['.mp4', '.webm', '.mkv', '.mov', '.avi', '.flv'],
    mimeTypes: ['video/mp4', 'video/webm', 'video/x-matroska', 'video/quicktime', 'video/x-msvideo', 'video/x-flv'],
    maxSize: 2000 * 1024 * 1024, // 2GB
  },
  audio: {
    extensions: ['.mp3', '.wav', '.ogg', '.m4a', '.opus', '.oga', '.flac'],
    mimeTypes: ['audio/mpeg', 'audio/wav', 'audio/x-wav', 'audio/ogg', 'audio/opus', 'audio/mp4', 'audio/x-m4a', 'audio/flac'],
    maxSize: 200 * 1024 * 1024, // 200MB
  },
  document: {
    extensions: ['.pdf', '.pptx', '.docx', '.doc', '.ppt', '.odt', '.rtf', '.epub', '.mobi'],
    mimeTypes: [
      'application/pdf',
      'application/vnd.openxmlformats-officedocument.presentationml.presentation',
      'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
      'application/msword',
      'application/vnd.ms-powerpoint',
      'application/vnd.oasis.opendocument.text',
      'application/rtf',
      'application/epub+zip',
      'application/x-mobipocket-ebook',
    ],
    maxSize: 200 * 1024 * 1024, // 200MB
  },
  spreadsheet: {
    extensions: ['.csv', '.xls', '.xlsx', '.ods'],
    mimeTypes: [
      'text/csv',
      'application/vnd.ms-excel',
      'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
      'application/vnd.oasis.opendocument.spreadsheet',
    ],
    maxSize: 100 * 1024 * 1024, // 100MB
  },
  archive: {
    extensions: ['.zip', '.rar', '.7z', '.tar', '.gz'],
    mimeTypes: [
      'application/zip',
      'application/x-zip-compressed',
      'application/x-rar-compressed',
      'application/vnd.rar',
      'application/x-7z-compressed',
      'application/x-tar',
      'application/gzip',
      'application/x-gzip',
    ],
    maxSize: 1000 * 1024 * 1024, // 1GB
  },
  text: {
    extensions: ['.txt', '.md', '.json', '.srt', '.vtt', '.py', '.js', '.ts', '.css', '.html', '.xml', '.cpp', '.c', '.java'],
    mimeTypes: [
      'text/plain',
      'text/markdown',
      'application/json',
      'text/vtt',
      'application/x-subrip',
      'text/x-python',
      'text/javascript',
      'text/typescript',
      'text/css',
      'text/html',
      'application/xml',
      'text/x-c++src',
      'text/x-csrc',
      'text/x-java-source',
    ],
    maxSize: 20 * 1024 * 1024, // 20MB
  },
} as const;

export type FileType = keyof typeof FILE_TYPES;

/**
 * Validate a file against allowed types and size limits
 */
export function validateFile(
  file: File,
  allowedTypes: FileType[],
  customMaxSize?: number,
): { valid: boolean; error?: string } {
  if (!file) {
    return { valid: false, error: 'No file selected' };
  }

  // Block SVG files explicitly for security
  if (file.name.toLowerCase().endsWith('.svg') || file.type === 'image/svg+xml') {
    return { valid: false, error: 'SVG files are not allowed for security reasons' };
  }

  // Find matching file type by MIME
  let matchedType: FileType | null = null;
  for (const type of allowedTypes) {
    const config = FILE_TYPES[type];
    if ((config.mimeTypes as readonly string[]).includes(file.type)) {
      matchedType = type;
      break;
    }
  }

  // Fallback to extension-only matching when MIME is unavailable or generic
  if (!matchedType) {
    const fileExtension = file.name.toLowerCase().split('.').pop();
    if (fileExtension) {
      for (const type of allowedTypes) {
        const config = FILE_TYPES[type] as { extensions: readonly string[] };
        if (config.extensions.includes(`.${fileExtension}`)) {
          matchedType = type;
          break;
        }
      }
    }
  }

  if (!matchedType) {
    const allowedMimes = allowedTypes.flatMap((type) => FILE_TYPES[type].mimeTypes).map(getFriendlyMimeName);
    return {
      valid: false,
      error: `Invalid file type: ${file.type || file.name}. Allowed types: ${allowedMimes.join(', ')}`,
    };
  }

  // Check file size
  const maxSize = customMaxSize || FILE_TYPES[matchedType].maxSize;
  if (file.size > maxSize) {
    const sizeMB = (file.size / 1024 / 1024).toFixed(1);
    const maxSizeMB = (maxSize / 1024 / 1024).toFixed(1);
    return {
      valid: false,
      error: `File too large (${sizeMB}MB). Maximum size: ${maxSizeMB}MB`,
    };
  }

  return { valid: true };
}

/**
 * Get accept attribute value for file inputs
 */
export function getAcceptValue(allowedTypes: FileType[]): string {
  return allowedTypes.flatMap((type) => FILE_TYPES[type].mimeTypes).join(',');
}

/**
 * Get human-readable description of allowed file types
 */
export function getFileTypeDescription(allowedTypes: FileType[]): string {
  const extensions = allowedTypes
    .flatMap((type) => FILE_TYPES[type].extensions)
    .map((ext) => ext.toUpperCase().slice(1))
    .join(', ');

  const maxSizes = [...new Set(allowedTypes.map((type) => FILE_TYPES[type].maxSize))];
  // Safely read the first/max size
  const [onlyMaxSize] = maxSizes;
  // Guard against undefined - TypeScript can't infer that maxSizes[0] exists even when length === 1
  const maxSizeStr =
    maxSizes.length === 1 && typeof onlyMaxSize !== 'undefined' ? `${onlyMaxSize / 1024 / 1024}MB` : 'varies';

  return `${extensions} (max ${maxSizeStr})`;
}

/**
 * Mapping of MIME types to short, user-friendly names
 */
export const MIME_TO_FRIENDLY_NAME: Record<string, string> = {
  'application/pdf': 'PDF',
  'image/jpeg': 'JPEG',
  'image/jpg': 'JPG',
  'image/png': 'PNG',
  'image/webp': 'WEBP',
  'image/gif': 'GIF',
  'image/avif': 'AVIF',
  'application/msword': 'DOC',
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document': 'DOCX',
  'application/vnd.oasis.opendocument.text': 'ODT',
  'text/csv': 'CSV',
  'application/vnd.ms-excel': 'XLS',
  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet': 'XLSX',
  'application/vnd.oasis.opendocument.spreadsheet': 'ODS',
  'application/vnd.ms-powerpoint': 'PPT',
  'application/vnd.openxmlformats-officedocument.presentationml.presentation': 'PPTX',
  'application/zip': 'ZIP',
  'application/x-zip-compressed': 'ZIP',
  'application/x-rar-compressed': 'RAR',
  'application/vnd.rar': 'RAR',
  'application/x-7z-compressed': '7Z',
  'application/x-tar': 'TAR',
  'application/gzip': 'GZ',
  'application/x-gzip': 'GZ',
  'text/plain': 'TXT',
  'text/markdown': 'MD',
  'application/json': 'JSON',
  'video/mp4': 'MP4',
  'video/webm': 'WEBM',
  'video/x-matroska': 'MKV',
  'video/quicktime': 'MOV',
  'audio/mpeg': 'MP3',
  'audio/wav': 'WAV',
  'audio/flac': 'FLAC',
  'application/epub+zip': 'EPUB',
  'application/x-mobipocket-ebook': 'MOBI',
  'text/x-python': 'PY',
  'text/javascript': 'JS',
  'text/typescript': 'TS',
  'text/css': 'CSS',
  'text/html': 'HTML',
  'application/xml': 'XML',
  'text/x-c++src': 'CPP',
  'text/x-csrc': 'C',
  'text/x-java-source': 'JAVA',
  'image/*': 'Images',
  'video/*': 'Videos',
  'audio/*': 'Audio',
};

/**
 * Returns a short, friendly name for a MIME type, or the MIME type itself if unknown.
 */
export function getFriendlyMimeName(mime: string): string {
  return MIME_TO_FRIENDLY_NAME[mime] ?? mime;
}
